File: Java/_698.java
/*
Greedy algo: 
Sort []nums, check if adding largest unused num would
fill subset exactly, and if yes, continue onto next 
subset. If adding largest unused would overflow subset, 
backtrack, then try next largest unused num. Continue 
until all bins are filled to goal ( == sum(nums)/ k)
*/

class Solution {
	List<Boolean> used;

    public boolean canPartitionKSubsets(int[] nums, int k) {
    	if (k == 1) return true;

        int goal = Arrays.stream(nums).sum();
        if (goal % k != 0) return false;

        used = new ArrayList<Boolean>(Arrays.asList(new Boolean[nums.length]));
        Collections.fill(used, false);

        goal /= k;

        nums = Arrays.stream(nums).boxed().sorted(Comparator.reverseOrder())
        .mapToInt(Integer::intValue)
        .toArray();
        
        return tryPartition(nums, k, goal, 0, 0);
    }

    private boolean tryPartition(int[] nums, int k, int goal, int curSum, int iToCheck) {
    	if (k == 1) return true;

    	if (curSum == goal) return tryPartition(nums, k - 1, goal, 0, 0);

    	for (int i = iToCheck; i < nums.length; i++) {
    		if (i > 0 && !used.get(i - 1) && nums[i - 1] == nums[i]) continue;

    		if (!used.get(i) && (curSum + nums[i]) <= goal) {
    			used.set(i, true);

    			if (tryPartition(nums, k, goal, curSum + nums[i], i + 1)) {
    				return true;
    			}
    			
    			used.set(i, false);
    		}
    	}
    	return false;
    }
}

File: Cpp/_698.cpp
class Solution {
public:
    bool* used;
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        if (k == 1) return true;
    
        int goal = std::accumulate(nums.begin(), nums.end(), 0);
        if (goal % k != 0) return false;
        
        goal /= k;
        
        used = new bool[nums.size()];
        std::fill(used, used + nums.size(), false);
        
        std::sort(nums.begin(), nums.end(), std::greater<>());
        
        return tryPartition(nums, k, goal);
    }
    
    bool tryPartition(const vector<int>& nums, int k, int goal,
                      int currSum = 0, int iToCheck = 0) {
        
        if (k == 1) return true;
        if (currSum == goal) return tryPartition(nums, k - 1, goal);
        
        for (int i = iToCheck; i < nums.size(); i++) {
            if (i > 0 && !used[i - 1] && nums[i - 1] == nums[i]) continue;
            
            if (!used[i] && (currSum + nums[i]) <= goal) {
                used[i] = true;
                
                if (tryPartition(nums, k, goal, currSum + nums[i], i + 1)) {
                    return true;
                }
                
                used[i] = false;
            }
        }
                    
        return false;
    }
};